src/server/game/Entities/Player/CollectionMgr.cpp

```diff
@@ -76,3 +76,209 @@ void CollectionMgr::ToySetFavorite(uint32 itemId, bool favorite)
+{
+    PreparedStatement* stmt = nullptr;
+    for (auto const& heirloom : _heirlooms)
+    {
+        stmt = LoginDatabase.GetPreparedStatement(LOGIN_REP_ACCOUNT_HEIRLOOMS);
+        stmt->setUInt32(0, _owner->GetBattlenetAccountId());
+        stmt->setUInt32(1, heirloom.first);
+        stmt->setUInt32(2, heirloom.second.flags);
+        trans->Append(stmt);
+    }
+}
+
+bool CollectionMgr::UpdateAccountHeirlooms(uint32 itemId, uint32 flags)
+{
+    return _heirlooms.insert(HeirloomContainer::value_type(itemId, HeirloomData(flags, 0))).second;
+}
+
+uint32 CollectionMgr::GetHeirloomBonus(uint32 itemId) const
+{
+    HeirloomContainer::const_iterator itr = _heirlooms.find(itemId);
+    if (itr != _heirlooms.end())
+        return itr->second.bonusId;
+
+    return 0;
+}
+
+void CollectionMgr::LoadHeirlooms()
+{
+    for (auto const& item : _heirlooms)
+    {
+        _owner->GetPlayer()->AddDynamicValue(PLAYER_DYNAMIC_FIELD_HEIRLOOMS, item.first);
+        _owner->GetPlayer()->AddDynamicValue(PLAYER_DYNAMIC_FIELD_HEIRLOOM_FLAGS, item.second.flags);
+    }
+}
+
+void CollectionMgr::AddHeirloom(uint32 itemId, uint32 flags)
+{
+    if (UpdateAccountHeirlooms(itemId, flags))
+    {
+        _owner->GetPlayer()->AddDynamicValue(PLAYER_DYNAMIC_FIELD_HEIRLOOMS, itemId);
+        _owner->GetPlayer()->AddDynamicValue(PLAYER_DYNAMIC_FIELD_HEIRLOOM_FLAGS, flags);
+    }
+}
+
+void CollectionMgr::UpgradeHeirloom(uint32 itemId, uint32 castItem)
+{
+    Player* player = _owner->GetPlayer();
+    if (!player)
+        return;
+
+    HeirloomEntry const* heirloom = sDB2Manager.GetHeirloomByItemId(itemId);
+    if (!heirloom)
+        return;
+
+    HeirloomContainer::iterator itr = _heirlooms.find(itemId);
+    if (itr == _heirlooms.end())
+        return;
+
+    uint32 flags = itr->second.flags;
+    uint32 bonusId = 0;
+
+    if (heirloom->UpgradeItemID[0] == castItem)
+    {
+        flags |= HEIRLOOM_FLAG_BONUS_LEVEL_90;
+        bonusId = heirloom->ItemBonusListID[0];
+    }
+    if (heirloom->UpgradeItemID[1] == castItem)
+    {
+        flags |= HEIRLOOM_FLAG_BONUS_LEVEL_100;
+        bonusId = heirloom->ItemBonusListID[1];
+    }
+
+    for (Item* item : player->GetItemListByEntry(itemId, true))
+        item->AddBonuses(bonusId);
+
+    // Get heirloom offset to update only one part of dynamic field
+    std::vector<uint32> const& fields = player->GetDynamicValues(PLAYER_DYNAMIC_FIELD_HEIRLOOMS);
+    uint8 offset = std::find(fields.begin(), fields.end(), itemId) - fields.begin();
+
+    player->SetDynamicValue(PLAYER_DYNAMIC_FIELD_HEIRLOOM_FLAGS, offset, flags);
+    itr->second.flags = flags;
+    itr->second.bonusId = bonusId;
+}
+
+void CollectionMgr::CheckHeirloomUpgrades(Item* item)
+{
+    Player* player = _owner->GetPlayer();
+    if (!player)
+        return;
+    
+    // Convert old heirloom to new ones
+    if (HeirloomEntry const* heirloom = sDB2Manager.GetHeirloomByOldItem(item->GetEntry()))
```

Comment:
This conversion should be done on database (should be a simple matter of updating entry).
This method is called in _LoadInventory, BEFORE bagslot and slot are set on the item - the returned position for DestroyItem is always bag 255 (main inventory, including equipped slots) slot 0 which is the currently equipped helm!